crypto/thread/api.c

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <openssl/configuration.h>
+#include <openssl/thread.h>
+#include <internal/thread.h>
+
+uint32_t OSSL_get_thread_support_flags(void)
+{
+    int support = 0;
+
+#if defined(OPENSSL_THREAD_POOL)
+    support |= OSSL_THREAD_SUPPORT_FLAG_THREAD_POOL;
+#endif
+#if defined(OPENSSL_DEFAULT_THREAD_POOL)
+    support |= OSSL_THREAD_SUPPORT_FLAG_DEFAULT_SPAWN;
+#endif
+
+    return support;
+}
+
+#if !defined(OPENSSL_THREAD_POOL) && !defined(OPENSSL_DEFAULT_THREAD_POOL)
+
+int OSSL_set_max_threads(uint64_t max_threads)
+{
+    return 0;
+}
+
+uint64_t OSSL_get_max_threads(void)
+{
+    return 0;
+}
+
+#else
+
+uint64_t OSSL_get_max_threads(void)
+{
+    uint64_t ret = 0;
+    OSSL_LIB_CTX_THREADS *tdata = OSSL_LIB_CTX_GET_THREADS(NULL);
+
+    if (tdata == NULL)
+        goto fail;
+
+    ossl_crypto_mutex_lock(tdata->lock);
+    ret = tdata->max_threads;
+    ossl_crypto_mutex_unlock(tdata->lock);
+
+fail:
+    return ret;
+}
+
+int OSSL_set_max_threads(uint64_t max_threads)
+{
+    OSSL_LIB_CTX_THREADS *tdata;
+
+    tdata = OSSL_LIB_CTX_GET_THREADS(NULL);
```

Comment:
Shouldn't we be passing a libctx here?

Reply:
The thread pool design document omitted libctx from the API, namely from the OSSL_{get,set}_max_threads functions. I assumed that only meant that threads should use single global context.

I am not aware of any other way of getting the context in a general setting (for example, outside of a provider), so for the time being I added a formal ctx parameter to the two functions mentioned.

Is there a better way of getting libctx or shall we change the design doc?

Reply:
I think passing a libctx here is probably the way to go. However if we don't do that then I wouldn't get this data out of the default libctx. Instead I would just store it in global variables and initialise a lock using `RUN_ONCE`.

Reply:
Since you have already written the code to get this data out of a libctx, I would modify these functions to take a libctx for now while the discussion in #18864 goes to OTC. If OTC decides not to add the libctx then only then I would modify this to use a global variable/`RUN_ONCE`.